src/compiler/emitter.ts

```diff
@@ -168,155 +167,60 @@ module ts {
-                }
-
-                if (name.length === 2 && name.charCodeAt(1) >= CharacterCodes.a && name.charCodeAt(1) <= CharacterCodes.z) {
-                    // handles _a .. _z
-                    let n = name.charCodeAt(1) - CharacterCodes.a;
-                    return n < tempCount;
-                }
-                else {
-                    // handles _1, _2...
-                    let n = +name.substring(1);
-                    return !isNaN(n) && n >= 0 && n < (tempCount - 26);
-                }
-            }
-
-            // This function generates a name using the following pattern:
-            // _a .. _h, _j ... _z, _0, _1, ...
-            // It is guaranteed that generated name will not shadow any existing user-defined names,
-            // however it can hide another name generated by this function higher in the scope.
-            // NOTE: names generated by 'makeTempVariableName' and 'makeUniqueName' will never conflict.
-            // see comment for 'makeTempVariableName' for more information.
-            function makeTempVariableName(location: Node, tempVariableKind: TempVariableKind): string {
-                let tempName: string;
-                if (tempVariableKind !== TempVariableKind.auto && !(predefinedTempsInUse & tempVariableKind)) {
-                    tempName = tempVariableKind === TempVariableKind._i ? "_i" : "_n";
-                    if (!resolver.resolvesToSomeValue(location, tempName)) {
-                        predefinedTempsInUse |= tempVariableKind;
-                        return tempName;
+            // Return the next available name in the pattern _a ... _z, _0, _1, ...
+            // TempFlags._i or TempFlags._n may be used to express a preference for that dedicated name.
+            // Note that names generated by makeTempVariableName and makeUniqueName will never conflict.
+            function makeTempVariableName(flags: TempFlags): string {
+                if (flags && !(tempFlags & flags)) {
+                    var name = flags === TempFlags._i ? "_i" : "_n"
+                    if (isUniqueName(name)) {
+                        tempFlags |= flags;
+                        return name;
                     }
                 }
-
-                do {
-                    // Note: we avoid generating _i and _n as those are common names we want in other places.
-                    var char = CharacterCodes.a + tempCount;
-                    if (char !== CharacterCodes.i && char !== CharacterCodes.n) {
-                        if (tempCount < 26) {
-                            tempName = "_" + String.fromCharCode(char);
-                        }
-                        else {
-                            tempName = "_" + (tempCount - 26);
-                        }
+                while (true) {
+                    let count = tempFlags & TempFlags.CountMask;
+                    let ch = CharacterCodes.a + count;

```

Comment:
this `ch` will be meaningless if `count >= 26`. I would not even compute it in that case, because it will be distracting.